src/Core/Checkout/Payment/PaymentProcessor.php

```diff
@@ -111,6 +111,22 @@ public function finalize(TokenStruct $token, Request $request, SalesChannelConte
 
         try {
             $transactionStruct = $this->paymentTransactionStructFactory->build($token->getTransactionId(), $context->getContext());
+
+            if ($token->isInvalidated()) {
+                // Token was already handled - check current state of the transaction
+                $stateName = $transactionStruct->getOrderTransaction()->getStateMachineState()->getTechnicalName();
+                if ($stateName === OrderTransactionStates::STATE_AUTHORIZED || $stateName === OrderTransactionStates::STATE_PAID || $stateName === OrderTransactionStates::STATE_PARTIALLY_PAID) {
+                    return $token;
+                }
+
+                if ($stateName === OrderTransactionStates::STATE_FAILED || $stateName === OrderTransactionStates::STATE_CANCELLED) {
+                    $token->setException(new TokenInvalidatedException($token->getToken() ?? ''));
+                    return $token;
+                }
```

Comment:
Note the description of #12593:
> [...]
> A possible solution was implemented in https://github.com/shopware/shopware/pull/4952, but it should be evaluated if relying on the transaction state is a good idea, as the payment method has to transition to a "safe" state themselves.
> ### Refined solution
> 
> We want to flag the token as being used to differentiate between an invalid and a used token. Upon receiving a used token, we want to skip and return.

Reply:
For now, in await of a better approach I've ported this